parsons/action_builder/action_builder.py

```diff
@@ -0,0 +1,351 @@
+                The list of strings of unique identifiers for a record being updated. ID strings
+                will need to begin with the origin system, followed by a colon, e.g.
+                `action_builder:abc123-...`.
+            data: dict
+                The details to include on the record being upserted, to be included as the value
+                of the `person` key. See
+                [documentation for the Person Signup Helper](https://www.actionbuilder.org/docs/v1/person_signup_helper.html#post)
+                for examples, and
+                [the Person endpoint](https://www.actionbuilder.org/docs/v1/people.html#field-names)
+                for full entity object composition.
+            campaign: str
+                Optional. The 36-character "interact ID" of the campaign whose data is to be
+                retrieved or edited. Not necessary if supplied when instantiating the class.
+        `Returns:`
+            Dict containing Action Builder entity data.
+        """  # noqa: E501
+
+        # Check that we have appropriate entity type/identifier, name, and campaign first
+        if entity_type == identifiers is None:
+            error_msg = "Must provide either entity_type (to insert a new record) "
+            error_msg += "or identifiers (to update an existing record)"
+            raise ValueError(error_msg)
+
+        if not isinstance(data, dict):
+            data = {}
+
+        name_check = [
+            key for key in data.get("person", {}) if key in ("name", "given_name")
+        ]
+        if identifiers is None and not name_check:
+            raise ValueError("Must provide name or given name if inserting new record")
+
+        campaign = self._campaign_check(campaign)
+
+        url = f"campaigns/{campaign}/people"
+
+        if "person" not in data:
+            # The POST data must live inside of person key
+            data["person"] = {}
+
+        if identifiers:
+            # Updating an existing record
+
+            if isinstance(identifiers, str):
+                # Ensure identifiers is a list, even if it will usually just be one item
+                identifiers = [identifiers]
+
+            # Default to assuming identifier comes from Action Builder and add prefix if missing
+            identifiers = [
+                f"action_builder:{x}" if ":" not in x else x for x in identifiers
+            ]
+            data["person"]["identifiers"] = identifiers
+
+        if entity_type:
+            # Inserting a new record
+            data["person"]["action_builder:entity_type"] = entity_type
+
+        return self.api.post_request(url=url, data=json.dumps(data))
+
+    def add_tags_to_record(
+        self, identifiers, tag_name, tag_field, tag_section, campaign=None
+    ):
+        """
+        Add a tag (i.e. custom field value) to an existing entity record in Action Builder. The
+        tags, along with their category/field and section/group, must already exist (except for
```

Comment:
is there a reason not to create one now if they don't already exist?

Reply:
I can imagine use-cases where someone would want to prevent that in case of bad data, so I think we should leave leave that for the logic used by folks writing to this connector, personally.